File: lib/interpreter.cpp
#include "parser.h"

class Interpreter
{
public:
    Interpreter(ASTNode *root) : root(root) {}

    int interpret()
    {
        return evaluate(root);
    }

private:
    ASTNode *root;

    int evaluate(ASTNode *node)
    {
        if (auto binOp = dynamic_cast<BinaryOp *>(node))
        {
            int left = evaluate(binOp->left);
            int right = evaluate(binOp->right);
            switch (binOp->op)
            {
            case TokenType::PLUS:
                return left + right;
            case TokenType::MINUS:
                return left - right;
                // Adicione outras operações conforme necessário
            }
        }
        else if (auto num = dynamic_cast<Number *>(node))
        {
            return num->value; // Não precisa de conversão
        }
        return 0;
    }   
};

File: lib/interpreter.py
import pdb

class Interpreter:
    def __init__(self, parser):
        self.parser = parser

    def visit(self, node):
        method_name = 'visit_' + type(node).__name__
        print("Method visited: ", method_name)
        visitor = getattr(self, method_name, self.generic_visit)
        return visitor(node)

    def visit_BinaryOp(self, node):
        if node.op == 'PLUS':
            return self.visit(node.left) + self.visit(node.right)
        elif node.op == 'MINUS':
            return self.visit(node.left) - self.visit(node.right)

    def visit_Num(self, node):
        return node.value

    def generic_visit(self, node):
        raise Exception('No visit_{} method'.format(type(node).__name__))

    def interpret(self):
        tree = self.parser.expr()
        pdb.set_trace()
        return self.visit(tree)
